### ISSUE TITLE
React ComponentType not checked correctly in the presence of a library definition importing it.

### ISSUE THREAD

**Issue by philipnilsson** [facebook/flow#6414]
## Flow version 
`0.73.0`

## Unexpected behavior

The following program type checks with no errors, but should have a missing prop error

```
// @flow

import React, { type ComponentType } from 'react';

const X: ComponentType<{ foo: string }> = () => null;

export const test = <X />;
```

when I have a library definition containing the line

```
import type { Component, ComponentType } from 'react';
```

... in my `flow-typed` directory.

- Replacing `ComponentType` with `React$ComponentType` in the failing example causes the error to be reported correctly.

- Removing the import and using `React$ComponentType` in the libdef causes the failing example to be reported correctly (in my case, along with a few hundred new - seemingly valid - issues in my code base)

## Expected

Either of

 - Flow reports the missing prop error in the presence of this import
 - Flow reports the library defintion as invalid instead of silenty failing.

**Comment by fishythefish** [comment]
Can you supply a minimal example? I can't reproduce your issue on [Try Flow](https://flow.org/try/#0JYWwDg9gTgLgBAJQKYEMDGMA0cDecYCeYScAwhOBAHZJUwAqRJAvnAGZQVwDkUqG3ANwAoYWmoBneAA0AXGQqQadRsQA8eNhAjypUYFQDmcZgD44AXjgAKAJSXzVAK4AbFyOFIAHpFhxxVFL4SEFWatJwAPSmgkA). Are you saying the line in the library definition is silencing this error?

**Comment by philipnilsson** [comment]
@fishythefish I'm not sure it is possible to reproduce this on try flow. I can only repro this with a `.flowconfig` and with library file present in the `flow-typed` folder. To repro you can create three files

### ./index.js
```
// @flow

import React, { type ComponentType } from 'react';
const X: ComponentType<{ foo: string }> = () => null;
const test = <X />;
```

### ./flow-typed/test.js
```
import type { Component, ComponentType } from 'react';
```

### .flowconfig
```
[libs]
./flow-typed/.*
```

The error will be silenced as described on `flow 0.73.0`.

**Comment by gnprice** [comment]
This was fixed in v0.144 (with 5148a475c0fe77e7a0d5c561ff0ee73070fd5fb1), so this issue should now be closed.

The repros above now give the following error:
```
Cannot use an import statement at the toplevel of a library file. Import statements may only
appear inside a `declare module`. [toplevel-library-import]

     1│ import type { Component, ComponentType } from 'react';
```

Moreover, even when you do have such an import, Flow correctly gives an error at the line that's wrong (and this part may have been true before v0.144, too; not sure):
```js
Cannot create X element because property foo is missing in props [1] but exists in object
type [2]. [prop-missing]

 [2]  6│ const X: ComponentType<{| foo: string |}> = () => null;
      7│
      8│ // $ExpectError
 [1]  9│ const test = <X />;
```

**Comment by gnprice** [comment]
Oh, and in fact the fix was released a bit earlier, in v0.143.0; the commit was cherry-picked as d35f6dc6495d63d6a33fe18608c3ec962c4463b9.

Just before that fix was the one in fc121d6a18b84f88c697d667c2bec6cb7650c407, which explains how the issue led to a lack of type-checking:

    Raise all lib errors
    
    Summary:
    Before, we would collect the errors encountered in each lib file, right
    after the pass over the code of the library, but before merging all library
    contexts into a single "master" context, and optimizing it. This resulted
    in us missing a number of errors. See for example the previous couple
    of diffs. Crucially, we would miss a lot of `[cannot-resolve-name]`
    errors, which meant that these types would silently turn into `any` for
    every file in the repo.
    
    This diff changes this by collecting the errors from the master context
    after context optimization. It then filters suppressions and stores the
    errors together with the rest of the errors in the server environment.
    
    […]
